aten/src/ATen/native/cuda/SoftMax.cu

```diff
@@ -939,34 +939,39 @@ TORCH_IMPL_FUNC(softmax_backward_cuda_out)
 }
 
-Tensor masked_softmax_cuda(const Tensor& input, const Tensor& mask) {
-    TORCH_CHECK(mask.scalar_type() == ScalarType::Bool, "Mask should be a boolean tensor");
-    bool is_transformer_mask = (input.dim() == 4 && mask.dim() == 2 && input.size(0) == mask.size(0) && input.size(2) == mask.size(1) && input.size(3) == mask.size(1));
-    TORCH_CHECK(mask.sizes() == input.sizes() || is_transformer_mask, "Mask shape should match input");
-    // Always do masked softmax on last dim
-    int softmax_elements = input.size(input.dim() - 1);
-    // Persistent softmax only support softmax_elements <= 1024,
-    // Therefore once softmax_elements > 1024, we need to fallback to vanilla masked_softmax
-    Tensor output = at::empty_like(input, input.options());
-    // Fallback to a slower masked softmax solution
-    if (softmax_elements > 1024 || softmax_elements * input.element_size() > 4096 || !mask.is_contiguous()) {
-        AT_DISPATCH_FLOATING_TYPES_AND2(
-          ScalarType::Half,
-          ScalarType::BFloat16,
-          input.scalar_type(),
-          "masked_softmax",
-          [&] {
-            Tensor mask_not = mask.logical_not();
-            output = at::softmax(input.masked_fill(mask_not, -std::numeric_limits<scalar_t>::infinity()), -1);
-          });
-        return output;
-    }
-    int batch_count = input.numel() / softmax_elements;
-    int chunk_size = input.numel() / input.size(0);
-    if (is_transformer_mask) {
+Tensor masked_softmax_cuda(const Tensor& input_, int64_t dim, const Tensor& mask_) {
+  Tensor output = at::empty_like(input_, input_.options());
+  TORCH_CHECK(mask_.scalar_type() == ScalarType::Bool, "Mask should be a boolean tensor");
+  bool is_transformer_mask = (input_.dim() == 4 && mask_.dim() == 2 && input_.size(0) == mask_.size(0) && input_.size(2) == mask_.size(1) && input_.size(3) == mask_.size(1));
+  TORCH_CHECK(mask_.sizes() == input_.sizes() || is_transformer_mask, "Mask shape should match input");
+
+  auto input = input_.dim() == 0 ? input_.view(1) : input_;
+  auto mask = mask_.dim() == 0 ? mask_.view(1) : mask_;
+
+  int softmax_elements = input.size(dim);
+  // Persistent softmax only support softmax_elements <= 1024,
+  // Therefore once softmax_elements > 1024, we need to fallback to vanilla masked_softmax
+  // Fallback to a slower masked softmax solution
+  if (softmax_elements > 1024 || softmax_elements * input.element_size() > 4096 || !mask.is_contiguous() || dim < input.dim()-1) {
+    AT_DISPATCH_FLOATING_TYPES_AND2(
+      ScalarType::Half,
+      ScalarType::BFloat16,
+      input.scalar_type(),
+      "masked_softmax",
+      [&] {
+        Tensor mask_not = mask.logical_not();
+        output = at::softmax(input.masked_fill(mask_not, -std::numeric_limits<scalar_t>::infinity()), dim);
+        // masked_softmax should return 0 in the output even when falling back to vanilla softmax
+        output = output.masked_fill(mask_not, 0);
```

Comment:
In the fallback functions, need to `masked_fill` to 0 in case there are `nan`s that arise. Could also use `nan_to_mean`, but this seems more correct since it will only affect those in the mask in case there are other `nan`s